### PR TITLE
Rewrite guestbook example to use kube-dns instead of host:port

### LINKED ISSUE #5091: Frontend cannot connect to redis data-base in guestbook example.
It comes up in the new guestbook.sh e2e test (#3693). I'll try to reproduce manually.

```
<br /> <b>Fatal error</b>: Uncaught exception 'Predis\Connection\ConnectionException' with message 'Error while reading line from the server [tcp://10.0.211.28:6379]' in /vendor/predis/predis/lib/Predis/Connection/AbstractConnection.php:141 Stack trace:
#0 /vendor/predis/predis/lib/Predis/Connection/StreamConnection.php(208): Predis\Connection\AbstractConnection-&gt;onConnectionError('Error while rea...')
#1 /vendor/predis/predis/lib/Predis/Connection/AbstractConnection.php(130): Predis\Connection\StreamConnection-&gt;read()
#2 /vendor/predis/predis/lib/Predis/Connection/AbstractConnection.php(122): Predis\Connection\AbstractConnection-&gt;readResponse(Object(Predis\Command\StringGet))
#3 /vendor/predis/predis/lib/Predis/Client.php(246): Predis\Connection\AbstractConnection-&gt;executeCommand(Object(Predis\Command\StringGet))
#4 /vendor/predis/predis/lib/Predis/Client.php(228): Predis\Client-&gt;executeCommand(Object(Predis\Command\StringGet))
#5 /app/index.php(32): Predis\Client-&gt;__call('get', Array)
#6 /app/index.php(32): Predis\Client-&gt;get('messag in <b>/vendor/predis/predis/lib/Predis/Connection/AbstractConnection.php</b> on line <b>141</b><br />```

### PR DESCRIPTION
**Pull request by piosz**
This fixes #5091

### CODE REVIEW

**Comment by satnam6502** [comment]
Happy to take for review this unless @thockin thinks it's his turf.

File: examples/guestbook/README.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ The web front end interacts with the redis master via javascript redis API calls
 ### Step Zero: Prerequisites
 
 This example assumes that you have a basic understanding of kubernetes services and that you have forked the repository and [turned up a Kubernetes cluster](https://github.com/GoogleCloudPlatform/kubernetes#contents):
-This example requires a kubernetes cluster.  
+This example requires a kubernetes cluster.
```
Comment (satnam6502) [inline examples/guestbook/README.md]: How about a capital K i.e. `Kubernetes cluster`

**Comment by satnam6502** [comment]
For my benefit, please can you explain a little bit more about the motivation for the PR?
I think the issue you are trying to solve #5901 relates to frontend pods being created before the redis services have their information propagated?

If that is the case, why not launch all the services first and still use SkyDNS?
And then launch the controllers/pods?
That way the env vars or DNS name will be defined?

Sorry if I am missing something here.

**Comment by satnam6502** [comment]
Satnam briefed via out of band communication.

**Comment by thockin** [comment]
please don't block on me - I am out this week

**Comment by satnam6502** [comment]
I thought I merged this earlier. Perhaps I failed to confirm. Retrying.

**Comment by piosz** [comment]
Thanks for review!

**Comment by satnam6502** [comment]
Just PD failures on e2e so I think we're golden.

**Comment by eparis** [comment]
I wish this had never been committed.  Our most basic example broken for people who just set up kubernetes in it's most bare form. People now have to have a WORKING kubernetes with a WORKING dns pod before they can even start to use the very first example. @nzwulfin is right, kube doesn't COME with dns.  Sure GKE might come with dns, the salt scripts might give you dns, but it isn't a PART of kube and our most basic example shouldn't expect it to be there.

My understanding is that you started the guestbook pods before you started the right services.  I know that stinks and the ENV var has it's limitations (also why I liked the idea of explicitly declaring required services)

@bgrant0607 

I'd like to hear the argument why I shouldn't try to have this reverted...

**Comment by bgrant0607** [comment]
@eparis 
For quite some time we have been moving towards making DNS required and eliminating the links env vars, at least by default.

Are you questioning whether we'll eventually want DNS by default for all providers, or just pointing out that that's not the current state?

**Comment by satnam6502** [comment]
DNS is also required to make other cluster components work e.g. collecting logs. Rather than reverting to a baseline for examples that can't assume DNS why not instead put energy into providing DNS functionality for the other providers?

The is a fundamental tension right now between producing good idiomatic examples (and I think this PR is a good change in this regard) vs. trying to produce examples that work across all platforms.

**Comment by nzwulfin** [comment]
My original question was around the prerequisites changing in the guestbook example without clear signaling.  I know it's in the commit history, but troubleshooting on the outside (not Kubernetes contributors) took a while to track down to this PR.

The prerequisites are listed as  "working cluster" with a pointer to a directory full of getting started guides.  However, the guestbook explicitly requires DNS, which isn't provided in any of the guides I can see.  It may be implicit in GKE and therefor not listed in the GKE guide, but DNS isn't explicitly configured in any guide.

Other contributors will have a harder time replicating idiomatic examples for other platforms if assumptions aren't present.

**Comment by satnam6502** [comment]
That's a very good point. Perhaps we have to bit the bullet and have variants of guestbook with clearly stated pre-requisites?

**Comment by eparis** [comment]
I'll go away, but I pretty strongly disagree with the assertion that I should need to have things running and functioning on TOP of the cluster before the most basic example of hos to USE the cluster is possible.

Logging and metrics is another example of services on top of the cluster.

Really I guess that means if we are going to continue to pursue the DNS for services model it would mean to me we need to make an effort to make the DNS be part of the cluster rather than an addon (even if an addon G*E does with salt)

I guess I want to see skyDNS be merged into the kube-proxy and the kubelet automatically just set the kube-proxy as the dns server, or something like that, if DNS is our requirement....

**Comment by nzwulfin** [comment]
If the change to the example would break on the current environment, then adding that change to the list should work without multiple variants.

For example, moving to DNS based services from ENV based services would break on a non-DNS cluster.  Previous examples didn't need DNS, so that goes on the list.

The API move from v1beta1 to v1beta3 is another change (if there are non-backward compatible changes) that might need to be listed, but could be handled by listing the version used to create / test the guide.

**Comment by derekwaynecarr** [comment]
I think its reasonable for an example to assert that it requires a pod running on the kube-master to support DNS.  @eparis - i see your point, but I am not sure we ever identified a particular example as the most basic example.  Maybe we need an example that does not require DNS, but demonstrates why in example 2 you would want to add a DNS pod to provide more value to the cluster.

**Comment by eparis** [comment]
In the interest of being a pain in the butt, DNS is not just launching a pod.  You also have to configure all of the kubelets with knowledge of how that pod was launched.

Kubelet must be configured to use DNS service, DNS service needs kubelet to launch. Inherent inception type wackiness.

**Comment by derekwaynecarr** [comment]
@eparis - i see your config point.  i also am aware that many getting started guides are configuring kube incorrectly aside from DNS.

**Comment by nzwulfin** [comment]
@derekwaynecarr If I've read the docs right, there's a flow (if hard to find the right entry point) of 101 -> 201 -> guestbook in the walkthrough example folder.  However, they don't really build that way, and if that's not the flow then it should be changed.  Users get dumped into the examples directory without any clear direction.  A lot of the examples refer to the getting started guides having provided some set of functionality, but many of them don't list those assumptions.

Maybe some guidelines around what this community determines is a "functional cluster" and then have those be the starting points for guides and examples to be written around?

**Comment by satnam6502** [comment]
Trying to tease apart the various issues here:
* Kubernetes is designed to be modular with clearly defined and loosely coupled sub-components. DNS is not part of the 'intrinsics' and it should be perfectly reasonable to use a Kubernetes cluster without DNS and allow users to provide some other method of their choice for service discovery/composition/etc. as they see fit. This speaks to several peoples' views on this thread.
* DNS is an add-on... but a very weird add-on... it provides a mechanism through which pods can 'compose' and this very useful mechanism is likely to be used by many applications which makes DNS look more like an 'operating system service' rather than some 'optional add-on'. Other cluster level services (cluster level monitoring and logging) do not influence how you architect your application in the way that the DNS add-on does.
* Given DNS's special role it should perhaps not be a cluster add-on pod/service but instead be supported as a core part of Kubernetes? However, this then forces upon people aspects of networking which they may not want or pay the cost for?
I don't see us changing how we do DNS before the 1.0 release so in the short term I think we do need to sort out the guestbook documentation or produce variants that assume different amounts of the onion.
I don't think @eparis should go away -- I think he makes lots of good points.

**Comment by eparis** [comment]
@satnam6502 changes before 1.0 isn't practical. Couldn't agree more.

/me goes to look for the 2.0 roadmap page...

**Comment by bgrant0607** [comment]
The post-1.0 roadmap will be a hot topic after we're not so crazy-busy trying to wrap up 1.0.

**Comment by bgrant0607** [comment]
cc @RichieEscarez regarding example flow: see https://github.com/GoogleCloudPlatform/kubernetes/pull/5284#issuecomment-107712719

**Comment by bgrant0607** [comment]
Rather than just discussing these issues in a closed PR, we should file new issues for them.

**Comment by satnam6502** [comment]
Created issue #9107

**Pull request merged**